internal/graph/check.go

```diff
@@ -945,6 +953,499 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
 	}
 }
 
+// recursiveMatchUserUsersetInfo groups common parameters needed
+// for recursiveMatchUserUserset for convenience purpose.
+type recursiveMatchUserUsersetInfo struct {
+	typesys                     *typesystem.TypeSystem
+	ds                          storage.RelationshipTupleReader
+	allowedUserTypeRestrictions []*openfgav1.RelationReference
+	userToUsersetMapping        storage.SortedSet
+	maxConcurrentReads          int
+
+	// The following member are atomic/sync in anticipation
+	// that the algorithm will parallelize the lookup.
+	dsCount        *atomic.Uint32
+	visitedUserset *sync.Map
+}
+
+// recursiveMatchUserUsersetFunc defines a function that recursively evaluates whether objects' matches userToUsersetMapping.
+type recursiveMatchUserUsersetFunc func(ctx context.Context,
+	req *ResolveCheckRequest,
+	commonParameters *recursiveMatchUserUsersetInfo,
+	level int) (*ResolveCheckResponse, error)
+
+func parallelizeRecursiveMatchUserUserset(ctx context.Context,
+	usersetItems []string,
+	req *ResolveCheckRequest,
+	commonParameters *recursiveMatchUserUsersetInfo,
+	recursiveFunc recursiveMatchUserUsersetFunc,
+	level int,
+) (*ResolveCheckResponse, error) {
+	checkOutcomeChan := make(chan checkOutcome, commonParameters.maxConcurrentReads)
+
+	pool := concurrency.NewPool(ctx, commonParameters.maxConcurrentReads)
+	cancellableCtx, cancelFunc := context.WithCancel(ctx)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	defer func() {
+		cancelFunc()
+		wg.Wait()
+	}()
+
+	go func() {
+		defer func() {
+			_ = pool.Wait()
+			close(checkOutcomeChan)
+			wg.Done()
+		}()
+		for _, usersetItem := range usersetItems {
+			_, visited := commonParameters.visitedUserset.LoadOrStore(usersetItem, struct{}{})
+			if !visited {
+				newReq := req.clone()
+				newTupleKey := tuple.NewTupleKey(usersetItem, req.GetTupleKey().GetRelation(), req.GetTupleKey().GetUser())
+				newReq.TupleKey = newTupleKey
```

Comment:
When you did `newReq.TupleKey.Object = usersetItem` you got a race condition, did you not? I got that too :( We should fix this in a separate PR

Reply:
https://github.com/openfga/openfga/issues/1995